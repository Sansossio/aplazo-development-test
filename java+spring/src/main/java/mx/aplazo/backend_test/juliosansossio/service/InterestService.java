package mx.aplazo.backend_test.juliosansossio.service;

import java.util.List;

import mx.aplazo.backend_test.juliosansossio.domain.CalculateInterest;
import mx.aplazo.backend_test.juliosansossio.domain.CalculateInterestResponse;
import mx.aplazo.backend_test.juliosansossio.domain.Interest;

public interface InterestService {
  public List<Interest> list();

  public Interest save(Interest interest);

  public List<CalculateInterestResponse> calculate(CalculateInterest body);
}
