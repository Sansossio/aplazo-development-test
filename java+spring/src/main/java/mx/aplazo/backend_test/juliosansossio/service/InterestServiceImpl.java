package mx.aplazo.backend_test.juliosansossio.service;

import java.util.ArrayList;
import java.util.Date;
import java.util.List;

import javax.validation.Valid;
import javax.validation.constraints.NotNull;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;
import org.springframework.validation.annotation.Validated;

import mx.aplazo.backend_test.juliosansossio.domain.CalculateInterest;
import mx.aplazo.backend_test.juliosansossio.domain.CalculateInterestResponse;
import mx.aplazo.backend_test.juliosansossio.domain.Interest;
import mx.aplazo.backend_test.juliosansossio.repository.InterestRepository;

@Service
@Validated
public class InterestServiceImpl implements InterestService {
    @Autowired
    private InterestRepository interestRepository;

    public List<Interest> list() {
        return interestRepository.findAll(Sort.by(Sort.Order.desc("id")));
    }

    public Interest save(@NotNull @Valid Interest interest) {
        return interestRepository.save(interest);
    }

    public List<CalculateInterestResponse> calculate(@Valid CalculateInterest body) {
        List<CalculateInterestResponse> response = new ArrayList<>();

        Double amount = body.getAmount();
        Integer terms = body.getTerms();
        Double rate = body.getRate();

        Double totalToPay = amount + (rate / 100) * amount;
        Double totalPerPayment = totalToPay / terms;

        for (int i = 1; i <= terms; i++) {
            CalculateInterestResponse calculateInterestResponse = new CalculateInterestResponse();
            calculateInterestResponse.setPayment_number(i);
            calculateInterestResponse.setAmount(totalPerPayment);
            calculateInterestResponse.setPayment_date(new Date());
            response.add(calculateInterestResponse);
        }

        return response;
    }
}
