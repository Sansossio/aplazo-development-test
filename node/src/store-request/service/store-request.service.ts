import { Injectable } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository } from 'typeorm';
import { StoreRequestEntity } from '../../database/models/store-request.entity';

const LIMIT = 10;

@Injectable()
export class StoreRequestService {
  constructor(
    @InjectRepository(StoreRequestEntity)
    private readonly storeRequestRepository: Repository<StoreRequestEntity>,
  ) {}

  async storeEvent(body: Object, response: Object) {
    const storeRequest = this.storeRequestRepository.create({
      request: JSON.stringify(body),
      response: JSON.stringify(response),
    });
    await this.storeRequestRepository.save(storeRequest);
  }

  async getEvents() {
    return this.storeRequestRepository.find({
      take: LIMIT,
      order: {
        id: 'DESC',
      },
    });
  }
}
