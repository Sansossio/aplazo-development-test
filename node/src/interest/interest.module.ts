import { Module } from '@nestjs/common';
import { InterestController } from './controller/interest.controller';
import { InterestService } from './service/interest.service';

@Module({
  controllers: [InterestController],
  providers: [InterestService],
})
export class InterestModule {}
