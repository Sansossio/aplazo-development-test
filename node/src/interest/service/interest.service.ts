import { Injectable } from '@nestjs/common';
import { StoreRequest } from '../../store-request/decorator/store-request.decorator';
import { CalculateInterestRequestDto } from '../dto/calculate.request.interest.dto';
import { CalculateInterestResponseDto } from '../dto/calculate.response.interest.dto';

@Injectable()
export class InterestService {
  @StoreRequest()
  calculateInterest(
    data: CalculateInterestRequestDto,
  ): CalculateInterestResponseDto[] {
    const response: CalculateInterestResponseDto[] = [];
    const totalToPay = data.amount + (data.rate / 100) * data.amount;
    const totalPerPayment = totalToPay / data.terms;

    for (let i = 1; i <= data.terms; i++) {
      response.push({
        payment_number: i,
        amount: totalPerPayment,
        payment_date: new Date(),
      });
    }
    return response;
  }
}
