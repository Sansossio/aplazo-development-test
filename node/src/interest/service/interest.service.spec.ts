import { InterestService } from './interest.service';

describe('InterestService', () => {
  let service: InterestService;

  jest.useFakeTimers().setSystemTime(new Date('2022-04-04').getTime());

  beforeEach(async () => {
    service = new InterestService();
  });

  it('should be defined', () => {
    expect(service).toBeDefined();
  });

  it('should calculate interest', () => {
    const request = {
      amount: 1000,
      rate: 10,
      terms: 12,
    };
    const response = service.calculateInterest(request);
    expect(response).toBeDefined();
    expect(response.length).toBe(12);
  });

  it('should calculate decimal interest response', () => {
    const request = {
      amount: 1000,
      rate: 1.25,
      terms: 12,
    };
    expect(service.calculateInterest(request)).toMatchSnapshot();
  });

  it('should calculate exact interest response', () => {
    const request = {
      amount: 1000,
      rate: 10,
      terms: 10,
    };
    expect(service.calculateInterest(request)).toMatchSnapshot();
  });
});
