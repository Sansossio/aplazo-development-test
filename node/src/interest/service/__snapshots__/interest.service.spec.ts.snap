// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InterestService should calculate decimal interest response 1`] = `
Array [
  Object {
    "amount": 84.375,
    "payment_date": 2022-04-04T00:00:00.000Z,
    "payment_number": 1,
  },
  Object {
    "amount": 84.375,
    "payment_date": 2022-04-04T00:00:00.000Z,
    "payment_number": 2,
  },
  Object {
    "amount": 84.375,
    "payment_date": 2022-04-04T00:00:00.000Z,
    "payment_number": 3,
  },
  Object {
    "amount": 84.375,
    "payment_date": 2022-04-04T00:00:00.000Z,
    "payment_number": 4,
  },
  Object {
    "amount": 84.375,
    "payment_date": 2022-04-04T00:00:00.000Z,
    "payment_number": 5,
  },
  Object {
    "amount": 84.375,
    "payment_date": 2022-04-04T00:00:00.000Z,
    "payment_number": 6,
  },
  Object {
    "amount": 84.375,
    "payment_date": 2022-04-04T00:00:00.000Z,
    "payment_number": 7,
  },
  Object {
    "amount": 84.375,
    "payment_date": 2022-04-04T00:00:00.000Z,
    "payment_number": 8,
  },
  Object {
    "amount": 84.375,
    "payment_date": 2022-04-04T00:00:00.000Z,
    "payment_number": 9,
  },
  Object {
    "amount": 84.375,
    "payment_date": 2022-04-04T00:00:00.000Z,
    "payment_number": 10,
  },
  Object {
    "amount": 84.375,
    "payment_date": 2022-04-04T00:00:00.000Z,
    "payment_number": 11,
  },
  Object {
    "amount": 84.375,
    "payment_date": 2022-04-04T00:00:00.000Z,
    "payment_number": 12,
  },
]
`;

exports[`InterestService should calculate exact interest response 1`] = `
Array [
  Object {
    "amount": 110,
    "payment_date": 2022-04-04T00:00:00.000Z,
    "payment_number": 1,
  },
  Object {
    "amount": 110,
    "payment_date": 2022-04-04T00:00:00.000Z,
    "payment_number": 2,
  },
  Object {
    "amount": 110,
    "payment_date": 2022-04-04T00:00:00.000Z,
    "payment_number": 3,
  },
  Object {
    "amount": 110,
    "payment_date": 2022-04-04T00:00:00.000Z,
    "payment_number": 4,
  },
  Object {
    "amount": 110,
    "payment_date": 2022-04-04T00:00:00.000Z,
    "payment_number": 5,
  },
  Object {
    "amount": 110,
    "payment_date": 2022-04-04T00:00:00.000Z,
    "payment_number": 6,
  },
  Object {
    "amount": 110,
    "payment_date": 2022-04-04T00:00:00.000Z,
    "payment_number": 7,
  },
  Object {
    "amount": 110,
    "payment_date": 2022-04-04T00:00:00.000Z,
    "payment_number": 8,
  },
  Object {
    "amount": 110,
    "payment_date": 2022-04-04T00:00:00.000Z,
    "payment_number": 9,
  },
  Object {
    "amount": 110,
    "payment_date": 2022-04-04T00:00:00.000Z,
    "payment_number": 10,
  },
]
`;
